python/lsst/rubintv/production/redisUtils.py

```diff
@@ -346,93 +403,186 @@ def getNumFinished(self, instrument, taskName, processingId):
+        processingId = dataIdNoDetector.to_json().encode("utf-8")
         value = self.redis.hget(key, processingId)
         return int(value or 0)
 
-    def reportVisitLevelFinished(self, instrument, step, failed=False):
-        """Count the number of times a visit-level pipeline has finished.
+    def getAllDataIdsForTask(self, instrument: str, taskName: str) -> list[DataCoordinate]:
+        """Get a list of processed ids for the specified task
+
+        These are returned
+        """
+        key = f"{instrument}-{taskName}-FINISHEDCOUNTER"
+        idList = self.redis.hgetall(key).keys()  # list of bytes
+        return [
+            DataCoordinate.from_json(dataCoordJson, universe=self.butler.dimensions)
+            for dataCoordJson in idList
+        ]
+
+    def removeTaskCounter(self, instrument: str, taskName: str, dataId: DataCoordinate) -> None:
+        """Once a gather step is finished with all the expected data present,
+        remove the counter from the tracking dictionary to save reprocessing it
+        each time.
+
+        Parameters
+        ----------
+        instrument : `str`
+            The name of the instrument.
+        taskName : `str`
+            The name of the task that has finished processing.
+        dataId : `DataCoordinate`
+            The dataId the task has finished processing.
+        """
+        key = f"{instrument}-{taskName}-FINISHEDCOUNTER"
+        dataIdNoDetector = removeDetector(dataId, self.butler)
+        processingId = dataIdNoDetector.to_json().encode("utf-8")
+        if self.redis.hexists(key, processingId):
+            self.redis.hdel(key, processingId)
+        else:
+            self.log.warning(
+                f"Key {key} with processingId {processingId.decode('utf-8')} from {dataId} for {taskName}"
+                " did not exist when removal was attempted"
+            )
+
+    def reportDetectorLevelFinished(
+        self, instrument: str, step: str, who: str, processingId: str, failed=False
+    ) -> None:
+        """Count the number of times a detector-level pipeline has finished.
+
+        Increments the FINISHEDCOUNTER for the key corresponding to the
+        instrument, step, and who. If the processing failed, the FAILEDCOUNTER
+        is also incremented.
 
         Parameters
         ----------
         instrument : `str`
             The name of the instrument.
         step : `str`
             The name of the step which finished processing.
+        who : `str`
+            Who are we running the pipeline for, e.g. "SFM" or "AOS".
+        processingId : `str`
+            The processing id, either single or compound.
         failed : `bool`
             True if the processing did not fail to complete
         """
-        key = f"{instrument}-{step}-FINISHEDCOUNTER"
-        self.redis.incr(key, 1)  # creates the key if it doesn't exist
+        key = f"{instrument}-{step}-{who}-FINISHEDCOUNTER"
```

Comment:
Ah, interesting, OK. I can coordinate. And agree that this could do with a little wrapping/classing, it's been on my mind for a while to make it a little more codey and less arbitrary.